File: internal/service/flow_runtime_service.go
package service

import (
	"encoding/json"
	"errors"
	"fmt"
	"math/rand"

	"github.com/nexus-flow/nexus/internal/cache"
	"github.com/nexus-flow/nexus/internal/engine"
	"github.com/nexus-flow/nexus/internal/repository"
)

type FlowResult struct {
	FlowInstanceID string                 `json:"flowInstanceId"`
	Status         string                 `json:"status"`
	Data           map[string]interface{} `json:"data,omitempty"`
}

func TriggerFlow(flowKey, flowVersion string, flowData map[string]interface{}) (*FlowResult, error) {
	fv, err := repository.GetEnabledFlowVersion(flowKey, flowVersion)
	if err != nil {
		return nil, errors.New("流程版本不存在或未启用")
	}

	flowInfo, err := repository.GetFlowInfoByKey(flowKey)
	if err != nil {
		return nil, errors.New("流程不存在")
	}

	instanceID := fmt.Sprintf("%s_%s", flowInfo.FlowType, randomString(16))

	// Parse inputs/outputs/variables
	var inputs, outputs []engine.ParameterDef
	var variables []engine.VariableDef
	json.Unmarshal([]byte(fv.Inputs), &inputs)
	json.Unmarshal([]byte(fv.Outputs), &outputs)
	json.Unmarshal([]byte(fv.Variables), &variables)

	flowDef := &engine.FlowDef{
		InstanceID:  instanceID,
		FlowKey:     flowKey,
		FlowType:    flowInfo.FlowType,
		FlowContent: fv.FlowContent,
		Inputs:      inputs,
		Outputs:     outputs,
		Variables:   variables,
	}

	if flowInfo.FlowType == "async" {
		go func() {
			result := engine.Execute(flowDef, flowData)
			cache.GetCache().Set(instanceID, result)
		}()
		return &FlowResult{
			FlowInstanceID: instanceID,
			Status:         "RUNNING",
		}, nil
	}

	// sync
	result := engine.Execute(flowDef, flowData)
	return &FlowResult{
		FlowInstanceID: instanceID,
		Status:         result.Status,
		Data:           result.Data,
	}, nil
}

func TriggerFlowByDefinition(flowKey string, flowData map[string]interface{}) (*FlowResult, error) {
	fd, err := repository.GetFlowDefinitionByKey(flowKey)
	if err != nil {
		return nil, errors.New("流程定义不存在")
	}

	instanceID := fmt.Sprintf("debug_%s", randomString(16))

	// Load params from flow definition
	inputParams, _ := repository.ListParametersBySourceAndType("flow", fd.ID, 1)
	outputParams, _ := repository.ListParametersBySourceAndType("flow", fd.ID, 2)
	variableInfos, _ := repository.ListVariableInfosByFlowDefID(fd.ID)

	var inputs []engine.ParameterDef
	for _, p := range inputParams {
		inputs = append(inputs, engine.ParameterDef{Key: p.ParamKey, Name: p.ParamName, DataType: p.DataType})
	}
	var outputs []engine.ParameterDef
	for _, p := range outputParams {
		outputs = append(outputs, engine.ParameterDef{Key: p.ParamKey, Name: p.ParamName, DataType: p.DataType})
	}
	var variables []engine.VariableDef
	for _, v := range variableInfos {
		variables = append(variables, engine.VariableDef{Key: v.VariableKey, Name: v.VariableName, VariableType: v.VariableType, DataType: v.DataType})
	}

	flowDef := &engine.FlowDef{
		InstanceID:  instanceID,
		FlowKey:     flowKey,
		FlowType:    fd.FlowType,
		FlowContent: fd.FlowContent,
		Inputs:      inputs,
		Outputs:     outputs,
		Variables:   variables,
	}

	result := engine.Execute(flowDef, flowData)
	return &FlowResult{
		FlowInstanceID: instanceID,
		Status:         result.Status,
		Data:           result.Data,
	}, nil
}

func GetAsyncFlowResult(instanceID string) (map[string]interface{}, error) {
	result, ok := cache.GetCache().Get(instanceID)
	if !ok {
		return nil, errors.New("流程执行结果不存在或尚未完成")
	}
	if r, ok := result.(*engine.ExecutionResult); ok {
		return r.Data, nil
	}
	return nil, errors.New("结果格式异常")
}

func randomString(n int) string {
	const letters = "abcdefghijklmnopqrstuvwxyz0123456789"
	b := make([]byte, n)
	for i := range b {
		b[i] = letters[rand.Intn(len(letters))]
	}
	return string(b)
}

File: server/src/main/java/com/javaoffers/nexus/core/service/FlowRuntimeService.java
package com.javaoffers.nexus.core.service;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.javaoffers.nexus.core.cache.Cache;
import com.javaoffers.nexus.core.engine.ExecutionResult;
import com.javaoffers.nexus.core.engine.FlowDef;
import com.javaoffers.nexus.core.engine.FlowEngine;
import com.javaoffers.nexus.core.engine.ParameterDef;
import com.javaoffers.nexus.core.engine.VariableDef;
import com.javaoffers.nexus.core.model.FlowDefinition;
import com.javaoffers.nexus.core.model.FlowInfo;
import com.javaoffers.nexus.core.model.FlowVersion;
import com.javaoffers.nexus.core.model.Parameter;
import com.javaoffers.nexus.core.model.VariableInfo;
import com.javaoffers.nexus.core.repository.FlowDefinitionMapper;
import com.javaoffers.nexus.core.repository.FlowInfoMapper;
import com.javaoffers.nexus.core.repository.FlowVersionMapper;
import com.javaoffers.nexus.core.repository.ParameterMapper;
import com.javaoffers.nexus.core.repository.VariableInfoMapper;
import com.javaoffers.nexus.pkg.exception.BizException;
import lombok.Data;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/** 对应 Go core/service/flow_runtime_service.go。 */
@Service
public class FlowRuntimeService {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final FlowEngine flowEngine;
    private final FlowInfoMapper flowInfoMapper;
    private final FlowVersionMapper flowVersionMapper;
    private final FlowDefinitionMapper flowDefinitionMapper;
    private final ParameterMapper parameterMapper;
    private final VariableInfoMapper variableInfoMapper;
    private final Cache cache;

    public FlowRuntimeService(FlowEngine flowEngine,
                              FlowInfoMapper flowInfoMapper,
                              FlowVersionMapper flowVersionMapper,
                              FlowDefinitionMapper flowDefinitionMapper,
                              ParameterMapper parameterMapper,
                              VariableInfoMapper variableInfoMapper,
                              Cache cache) {
        this.flowEngine = flowEngine;
        this.flowInfoMapper = flowInfoMapper;
        this.flowVersionMapper = flowVersionMapper;
        this.flowDefinitionMapper = flowDefinitionMapper;
        this.parameterMapper = parameterMapper;
        this.variableInfoMapper = variableInfoMapper;
        this.cache = cache;
    }

    @Data
    public static class FlowResult {
        private String flowInstanceId;
        private String status;
        private Map<String, Object> data;
    }

    public FlowResult triggerFlow(String flowKey, String flowVersion, Map<String, Object> flowData) {
        FlowInfo flowInfo = flowInfoMapper.getByKey(flowKey);
        if (flowInfo == null) {
            throw new BizException("流程不存在");
        }
        FlowVersion fv = flowVersionMapper.getEnabledByFlowIdAndVersion(flowInfo.getId(), flowVersion);
        if (fv == null) {
            throw new BizException("流程版本不存在或未启用");
        }

        String instanceID = flowInfo.getFlowType() + "_" + randomString(16);

        List<ParameterDef> inputs = parseParams(fv.getInputs());
        List<ParameterDef> outputs = parseParams(fv.getOutputs());
        List<VariableDef> variables = parseVars(fv.getVariables());

        FlowDef flowDef = new FlowDef();
        flowDef.setInstanceId(instanceID);
        flowDef.setFlowKey(flowKey);
        flowDef.setFlowType(flowInfo.getFlowType());
        flowDef.setFlowContent(fv.getFlowContent());
        flowDef.setInputs(inputs);
        flowDef.setOutputs(outputs);
        flowDef.setVariables(variables);

        if ("async".equals(flowInfo.getFlowType())) {
            new Thread(() -> {
                ExecutionResult result = flowEngine.execute(flowDef, flowData);
                cache.set(instanceID, result);
            }).start();
            FlowResult r = new FlowResult();
            r.setFlowInstanceId(instanceID);
            r.setStatus("RUNNING");
            return r;
        }

        ExecutionResult result = flowEngine.execute(flowDef, flowData);
        FlowResult r = new FlowResult();
        r.setFlowInstanceId(instanceID);
        r.setStatus(result.getStatus());
        r.setData(result.getData());
        return r;
    }

    public FlowResult triggerFlowByDefinition(String flowKey, Map<String, Object> flowData) {
        FlowDefinition fd = flowDefinitionMapper.getByKey(flowKey);
        if (fd == null) {
            throw new BizException("流程定义不存在");
        }
        String instanceID = "debug_" + randomString(16);

        List<Parameter> inputParams = parameterMapper.listBySourceAndType("flow", fd.getId(), 1);
        List<Parameter> outputParams = parameterMapper.listBySourceAndType("flow", fd.getId(), 2);
        List<VariableInfo> variableInfos = variableInfoMapper.listByFlowDefId(fd.getId());

        List<ParameterDef> inputs = new ArrayList<>();
        if (inputParams != null) {
            for (Parameter p : inputParams) {
                ParameterDef pd = new ParameterDef();
                pd.setKey(p.getParamKey());
                pd.setName(p.getParamName());
                pd.setDataType(p.getDataType());
                inputs.add(pd);
            }
        }
        List<ParameterDef> outputs = new ArrayList<>();
        if (outputParams != null) {
            for (Parameter p : outputParams) {
                ParameterDef pd = new ParameterDef();
                pd.setKey(p.getParamKey());
                pd.setName(p.getParamName());
                pd.setDataType(p.getDataType());
                outputs.add(pd);
            }
        }
        List<VariableDef> variables = new ArrayList<>();
        if (variableInfos != null) {
            for (VariableInfo v : variableInfos) {
                VariableDef vd = new VariableDef();
                vd.setKey(v.getVariableKey());
                vd.setName(v.getVariableName());
                vd.setVariableType(v.getVariableType());
                vd.setDataType(v.getDataType());
                variables.add(vd);
            }
        }

        FlowDef flowDef = new FlowDef();
        flowDef.setInstanceId(instanceID);
        flowDef.setFlowKey(flowKey);
        flowDef.setFlowType(fd.getFlowType());
        flowDef.setFlowContent(fd.getFlowContent());
        flowDef.setInputs(inputs);
        flowDef.setOutputs(outputs);
        flowDef.setVariables(variables);

        ExecutionResult result = flowEngine.execute(flowDef, flowData);
        FlowResult r = new FlowResult();
        r.setFlowInstanceId(instanceID);
        r.setStatus(result.getStatus());
        r.setData(result.getData());
        return r;
    }

    public Map<String, Object> getAsyncFlowResult(String instanceID) {
        Object result = cache.get(instanceID);
        if (result == null) {
            throw new BizException("流程执行结果不存在或尚未完成");
        }
        if (result instanceof ExecutionResult) {
            return ((ExecutionResult) result).getData();
        }
        throw new BizException("结果格式异常");
    }

    private List<ParameterDef> parseParams(String json) {
        if (json == null || json.isEmpty()) {
            return new ArrayList<>();
        }
        try {
            return MAPPER.readValue(json, new TypeReference<List<ParameterDef>>() {});
        } catch (Exception e) {
            return new ArrayList<>();
        }
    }

    private List<VariableDef> parseVars(String json) {
        if (json == null || json.isEmpty()) {
            return new ArrayList<>();
        }
        try {
            return MAPPER.readValue(json, new TypeReference<List<VariableDef>>() {});
        } catch (Exception e) {
            return new ArrayList<>();
        }
    }

    private String randomString(int n) {
        String letters = "abcdefghijklmnopqrstuvwxyz0123456789";
        StringBuilder sb = new StringBuilder(n);
        for (int i = 0; i < n; i++) {
            sb.append(letters.charAt((int) (Math.random() * letters.length())));
        }
        return sb.toString();
    }
}
